### ISSUE TITLE
Is safe to remove value from list in asyncio

### ISSUE THREAD

**Issue by xiazhibin** [tiangolo/fastapi#5067]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class ConnectionManager:
    def __init__(self):
        self.active_connections: List[WebSocket] = []

    async def connect(self, websocket: WebSocket):
        await websocket.accept()
        self.active_connections.append(websocket)

    def disconnect(self, websocket: WebSocket):
        self.active_connections.remove(websocket)

    async def send_personal_message(self, message: str, websocket: WebSocket):
        await websocket.send_text(message)

    async def broadcast(self, message: str):
        for connection in self.active_connections:
            await connection.send_text(message)
```

### Description

I find the example from [doc](https://fastapi.tiangolo.com/advanced/websockets/?h=depends#handling-disconnections-and-multiple-clients).

I am wondering that it is safe to remove value from list.In this example, when websocketA broadcast message, `await websocketB.send_text(message)`, websocketB is disconnect,  what would happen?

### Operating System

Linux

### FastAPI Version

0.78.0

### Python Version

python3.8.13

**Comment by JarroVGIT** [comment]
In this case, all actions are run on the same event loop. No asyncio task (e.g. a coroutine, a Future object or an Future-like object) will be handling the list of `Websockets` while one of the coroutines is calling the `disconnect()` method. That is because, the method is a blocking (non-async) method. 

So, if request A is causing a `broadcast()` and request B is initating a disconnect by the user at the exact same time, only one of them will be executed at that moment and the other one will be waiting. 

The event loop changes which task is executed at which time, all the while some tasks are waiting for something to happen (e.g. `some_var = await something_that_returns_a_Future()`). 

So, for Starlette (and FastAPI by extension), this is perfectly safe to do. Hope this clarifies :)

**Comment by xiazhibin** [comment]
> Any line of code that does not include an `await` will execute serially with no other coroutines executing at the same time.
> 
> However, if you are using `run_in_executor` to turn synchronous/blocking code into async, that is _actually_ running in another thread and can cause data race conditions like you're describing.

Thanks for reply! So,I think the example from Doc is not rigorous enough

**Comment by JarroVGIT** [comment]
I would disagree with that statement. The purpose of the code block is to give an example on how to handle multiple connections and disconnects. It’s purpose is not to demonstrate how that should work in a threadsafe manner.

The code as-is, is perfectly safe to use. If you need multithreading or multiprocessing, then it is up to the developer to think about the ramifications and thing like race conditions. I don’t feel that should be very explicit in the examples though, as it just might confuse people more rather than helping the point of explaining how websockets work in FastAPI

**Comment by lucaswiman** [comment]
I was mentioning an edge case. What Jarro said is usually correct. The example code seems fine.
